### PR TITLE
internal/refcount.h: allow non-atomic build

### PR DESCRIPTION
**Pull request by levitte**
Configure with -DOPENSSL_DEV_NO_ATOMICS and you get refcount without
atomics.  This is intended for internal development only, to check the
refcounting is properly coded.  It should never become a configuration
option, hence the name of the macro.

### CODE REVIEW

**Comment by mspncp** [comment]
> This is intended for internal development only

Would it make sense to give the define a less official name then? Say, `OPENSSL_INTERNAL_NO_ATOMICS` (or whatever you prefer)?

**Comment by mspncp** [comment]
Or use `OPENSSL_DEV_NO_ATOMICS`.

**Comment by levitte** [comment]
> Or use `OPENSSL_DEV_NO_ATOMICS`.

I like that one.  Taken!

File: include/internal/refcount.h [hunk 1]
```diff
@@ -16,16 +16,17 @@
 #  endif
 # endif
 
-# if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L \
-     && !defined(__STDC_NO_ATOMICS__)
-#  include <stdatomic.h>
-#  define HAVE_C11_ATOMICS
-# endif
+# ifndef OPENSSL_DEV_NO_ATOMICS
```
Comment (mspncp) [inline include/internal/refcount.h]: I can't find the matching `#endif` ?

Comment (mspncp) [reply include/internal/refcount.h]: Oh, I found it :-)

File: include/internal/refcount.h [hunk 2]
```diff
@@ -112,9 +113,12 @@ static __inline int CRYPTO_DOWN_REF(volatile int *val, int *ret, void *lock)
     *ret = _InterlockedExchangeAdd(val, -1) - 1;
     return 1;
 }
+#   endif
+
 #  endif
+# endif
 
-# else
+# ifndef HAVE_ATOMICS
```
Comment (mspncp) [inline include/internal/refcount.h]: The most tricky part was to understand this change here, because it requires looking at all the entire ifdef-spagetti. But I think now that it's ok.

**Review by mspncp — DISMISSED** [review]
LGTM

File: include/internal/refcount.h (hunk 2 above)
Comment (levitte) [reply include/internal/refcount.h]: I've added a bit of commentary, I hope that helps

Comment (mspncp) [reply include/internal/refcount.h]: Yup!

Comment (mspncp) [reply include/internal/refcount.h]: BTW and FYI, just in case you ever had the same problem:

I found it hard to follow the preprocessor indentation using Emacs' incremental search, because by default it treats a single literal space character as 'any whitespace'. But I found a way to change that:

https://emacs.stackexchange.com/questions/2819/match-two-spaces-with-incremental-search

**Review by mspncp — APPROVED** [review]
I overlooked that you had dismissed my review. Reconfirmed.

(BTW: there is a new icon 'Re-request review' at the top right in the Reviewers list. I assume this will trigger a note.)

**Comment by levitte** [comment]
> (BTW: there is a new icon 'Re-request review' at the top right in the Reviewers list. I assume this will trigger a note.)

Ah!  Thanks

**Comment by levitte** [comment]
Merged.

503d4745a115b82db01c1fb22baaddb153d27cdb internal/refcount.h: allow non-atomic build
